pkg/meta/tkv_badger.go

```diff
@@ -0,0 +1,276 @@
+/*
+ * JuiceFS, Copyright 2021 Juicedata, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package meta
+
+import (
+	"bytes"
+	badger "github.com/dgraph-io/badger/v3"
+	"sync"
+)
+
+func init() {
+	Register("badger", newKVMeta)
+	drivers["badger"] = newBadgerClient
+}
+
+func newBadgerClient(addr string) (tkvClient, error) {
+	client, err := badger.Open(badger.DefaultOptions(addr))
+	if err != nil {
+		return nil, err
+	}
+	return &badgerClient{client, new(sync.RWMutex)}, err
+}
+
+type badgerTxn struct {
+	t *badger.Txn
+}
+
+func (tx *badgerTxn) scan(prefix []byte, handler func(key []byte, value []byte)) {
+	it := tx.t.NewIterator(badger.DefaultIteratorOptions)
+	defer it.Close()
+	for it.Seek(prefix); it.Valid(); it.Next() {
+		item := it.Item()
+		value, _ := item.ValueCopy(nil)
+		handler(it.Item().Key(), value)
+	}
+}
+
+func (tx *badgerTxn) get(key []byte) []byte {
+	item, err := tx.t.Get(key)
+	if err == badger.ErrKeyNotFound {
+		return nil
+	}
+	var value []byte
+	err = item.Value(func(val []byte) error {
+		value = append([]byte{}, val...)
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	return value
+}
+
+func (tx *badgerTxn) gets(keys ...[]byte) [][]byte {
+	values := make([][]byte, len(keys))
+	ret := make(map[string][]byte)
+	for _, key := range keys {
+		item, err := tx.t.Get(key)
+		if err == badger.ErrKeyNotFound {
+			ret[string(key)] = []byte{}
+		}
+		if err != nil {
+			panic(err)
+		}
+		err = item.Value(func(val []byte) error {
+			ret[string(key)] = append([]byte{}, val...)
```

Comment:
append may allocate more memory than make(), it's better to use `make([]byte)` here.